sdks/java/microbenchmarks/src/main/java/org/apache/beam/sdk/microbenchmarks/transforms/DoFnReflectorBenchmark.java

```diff
@@ -0,0 +1,239 @@
+import org.apache.beam.sdk.values.TupleTag;
+
+import org.joda.time.Instant;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+/**
+ * Benchmarks for {@link DoFn} and {@link DoFnWithContext} invocations, specifically
+ * for measuring the overhead of {@link DoFnReflector}.
+ */
+@State(Scope.Benchmark)
+@Fork(1)
+@Warmup(iterations = 5)
+public class DoFnReflectorBenchmark {
+
+  private static final String ELEMENT = "some string to use for testing";
+
+  private DoFn<String, String> doFn = new UpperCaseDoFn();
+  private DoFnWithContext<String, String> doFnWithContext = new UpperCaseDoFnWithContext();
+
+  private StubDoFnProcessContext stubDoFnContext = new StubDoFnProcessContext(doFn, ELEMENT);
+  private StubDoFnWithContextProcessContext stubDoFnWithContextContext =
+      new StubDoFnWithContextProcessContext(doFnWithContext, ELEMENT);
+  private ExtraContextFactory<String, String> extraContextFactory =
+      new ExtraContextFactory<String, String>() {
+
+    @Override
+    public BoundedWindow window() {
+      return null;
+    }
+
+    @Override
+    public WindowingInternals<String, String> windowingInternals() {
+      return null;
+    }
+  };
+
+  private DoFnReflector doFnReflector;
+  private DoFn<String, String> adaptedDoFnWithContext;
+
+  @Setup
+  public void setUp() {
+    doFnReflector = DoFnReflector.of(doFnWithContext.getClass());
+    adaptedDoFnWithContext = doFnReflector.toDoFn(doFnWithContext);
+  }
+
+  @Benchmark
+  public String invokeDoFn() throws Exception {
+    doFn.processElement(stubDoFnContext);
+    return stubDoFnContext.output;
+  }
+
+  @Benchmark
+  public String invokeDoFnWithContextViaAdaptor() throws Exception {
+    adaptedDoFnWithContext.processElement(stubDoFnContext);
+    return stubDoFnContext.output;
+  }
+
+  @Benchmark
+  public String invokeDoFnWithContext() throws Exception {
+    doFnReflector.invokeProcessElement(
+        doFnWithContext, stubDoFnWithContextContext, extraContextFactory);
+    return stubDoFnWithContextContext.output;
+  }
+
+  private static class UpperCaseDoFn extends DoFn<String, String> {
+
+    @Override
+    public void processElement(ProcessContext c) throws Exception {
+      c.output(c.element().toUpperCase());
+    }
+  }
+
+  private static class UpperCaseDoFnWithContext extends DoFnWithContext<String, String> {
+
+    @ProcessElement
+    public void processElement(ProcessContext c) throws Exception {
+      c.output(c.element().toUpperCase());
+    }
+  }
+
+  private static class StubDoFnProcessContext extends DoFn<String, String>.ProcessContext {
+
+    private final String element;
+    private String output;
+
+    public StubDoFnProcessContext(DoFn<String, String> fn, String element) {
+      fn.super();
```

Comment:
When instantiating a non-static inner class outside the context of an instance, it is necessary to wire it up to the enclosing instance. That is what this does.